New range display2

artkuo opened the pull request:
This is the result of a [user's forum discussion about ranges](https://groups.google.com/forum/#!searchin/julia-users/nicer/julia-users/qPqgJS-usrU/Hu40_tOlDQAJ). People were unhappy that `0:5` or `linspace(1,10,7)` does not return a vector, as they wanted. I believe it is sensible for these to be `range` objects (rather than being converted via `collect`), but the REPL display could be modified to show what the `range` actually looks like. The user probably wants to see some confirmation of what they were thinking, and seeing the definition doesn't help.

The proposed solution is to give REPL output like this:

```
julia> 0:4
5-element UnitRange{Int64}:
 0,1,2,3,4

julia> 0:100
101-element UnitRange{Int64}:
 0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,…,89,90,91,92,93,94,95,96,97,98,99,100

julia> linspace(0,2.5,4)
4-element LinSpace{Float64}:
 0.0,0.833333,1.66667,2.5
```

This is accomplished by overloading writemime (replutil.jl) to respond to `range` objects, adding a `print_range` method (range.jl) and appropriate unit tests (test/ranges.jl). This affects `display()`, but the user can still use `show(1:5)` to return the definition `1:5`. The `print_range` method is based on the existing `print_matrix_row`, and so it uses knowledge of screen size to make everything fit in one row, with horizontal dots (ellipsis) where appropriate.

Additional edits were made to show.jl, where I added a bunch of comments. I had to learn how `print_matrix` works, and the comments are meant to save time for someone like me, as the code is a bit opaque. There were absolutely no changes to execution of show.jl.

A few notes. I decided not to add brackets around the range. Perhaps square brackets would be appropriate, but I felt they were not needed, as a range is not exactly an array. Also, the range is printed as a single row to save vertical screen space. I thought this was acceptable because tuples are also printed across, despite acting somewhat like 1-d arrays, which are kind of "vertical." I would appreciate feedback if there are disagreements about these choices, or about the whole idea of modifying the REPL.

This is take 2: Fixed test to allow for 32- or 64-bit systems; changed pull request to base from master

andreasnoack committed:
Deprecate A_ldiv_B!(SparseMatrixCSC, StrideVecOrMat) thereby fixing #10787

Commit:
add same constructors as array

artkuo committed:
display ranges with their elements

artkuo committed:
added print_range for Range objects, to show the elements of the Range in REPL. Previous commits squashed, docstring moved back to front, builds and passes tests

artkuo committed:
Code comments for show.jl; no changes to execution code, only comments added while trying to figure out print_matrix and related functions.

artkuo committed:
make sure test works on either 32- or 64-bit systems (Int32 or Int64)

pao commented:
cf #13524

Commit:
add test for sharedarray constructor

artkuo committed:
Docstring comments for print_X methods in show. Comments are for un-exported methods such as alignment.

stevengj commented on base/range.jl:
```diff
@@ -244,6 +244,52 @@ function show(io::IO, r::LinSpace)
+#  1.0,2.0,3.0,…,4.0,5.0,6.0
+# This function is borrowed from print_matrix in show.jl
+# It is usually called by writemime (from replutil.jl)
+function print_range(io::IO, r::Union{Range,LinSpace},
```
This `Union` is unnecessary (similarly in `writemime`, below), because `LinSpace <: Range`.

stevengj commented on base/show.jl:
```diff
@@ -978,22 +983,28 @@ end
+# and figure out what's needed to be fully aligned, for example looking all
+# the way down a column and finding out the maximum size of each element.
+# Parameter sep is number of spaces to put between elements.
 function alignment(
```
The additional documentation is great, but you might as well put it in the form of a docstring:
``jl
"""
    alignment(X, rows, cols, cols_if_complete, cols_otherwise, sep)

Returns the alignment for the specified parts of array `X`, returning ....
...
"""
function alignment( ...
```

stevengj commented on test/ranges.jl:
```diff
@@ -544,6 +544,13 @@ for x in r
 @test i == 7
 
+# writemime should display the range or linspace nicely
+replstr(x) = sprint((io,x) -> writemime(io,MIME("text/plain"),x), x)
```
You can just use `replstr(x) = stringmime("text/plain", x)`

artkuo committed:
improved print_range, tweaked comments and tests

artkuo commented:
In response to comments, I switched writemime to stringmime in tests, removed the redundant Union{Range,LinSpace}, and changed the code comments I had added into Docstring format. 

Also, I discovered that Base.print_matrix is slow when displaying large matrices (try `eye(10^4)` in REPL), because it examines the spacing of the entire matrix, even when few elements will be printed. My code borrowed from print_matrix, but I have found a faster way and amended my print_range accordingly. This can be demonstrated by typing `0:10^9` at the REPL. I plan to go back and fix print_matrix as well and submit a separate pull request.

Thanks stevengj for suggesting helpful fixes.

artkuo committed:
streamlined writemime range type

stevengj commented on base/range.jl:
```diff
@@ -244,6 +244,72 @@ function show(io::IO, r::LinSpace)
+`pre` and `post` characters for each printed row, `sep` separator string between
+printed elements, `hdots` string for the horizontal ellipsis.
+"""
+"""
```
docstring is duplicated?

artkuo replied:
I wouldn't say duplicated, it's just that I'm not a huge fan of `f(a, [b, c])` to show optional parameters. I think it's easier for a reader to parse the simpler call first, and then see how additional parameters are added on. If the square brackets are considered the standard, I can switch to that, but otherwise I prefer my docstring.

stevengj replied:
Look again. You have the whole docstring twice.

artkuo replied:
doh! sorry about that; removed.

stevengj commented on test/ranges.jl:
```diff
@@ -544,6 +544,19 @@ for x in r
+# stringmime/writemime should display the range or linspace nicely
+# to test print_range in range.jl
+replstr(x) = stringmime("text/plain", x)
+@test replstr(1:4) == "4-element UnitRange{Int64}:\n 1,2,3,4" ||
```
Presumably you want:
```jl
@test replstr(1:4) == "4-element UnitRange{$Int}:\n 1,2,3,4"
```

artkuo replied:
Yes thanks, I've committed the change.

artkuo committed:
more concise tests

artkuo committed:
removed unwanted dup of docstring

stevengj commented:
Hmm, didn't it used to pass the tests?  What changed?

artkuo committed:
moved docstring below, due to strange problem with compilation

artkuo commented:
This is very strange. I deleted the extra docstring, and suddenly it wouldn't compile. I was able to move the docstring below the function definition, and it successfully makes and tests clean. I do not understand why it would work with a double docstring, and not with a single. I have no idea whether the bug was mine or elsewhere, but except for the odd location of the docstring, everything seems to work now.

stevengj commented on base/range.jl:
```diff
@@ -244,6 +244,75 @@ function show(io::IO, r::LinSpace)
     print(io, ')')
 end
 
+# `print_range(io, r)` prints out a nice looking range r in terms of its elements
```
don't duplicate the docstring in a comment.  Just say `# See docstring below.`

stevengj commented:
That's very odd.  @one-more-minute, could there be a parser/compiler bug here?  I don't see how duplicating the docstring could (correctly) change it from failure to success.

MichaelHatherly committed:
Fix doc bootstrap method error.

Docstrings defined before `==` is available for `Symbol` comparisons
causes a build error. Fix by comparing with `===`.

Ref: https://github.com/JuliaLang/julia/pull/13534#issuecomment-147906630

MichaelHatherly commented:
> suddenly it wouldn't compile.

@artkuo, was this the error you were getting?

```
LoadError(at "sysimg.jl" line 39: LoadError(at "range.jl" line 299: Base.MethodError(f===, args=(:function, :call))))
...
```

If so, I'll have a PR to fix this shortly.

>  I don't see how duplicating the docstring could (correctly) change it from failure to success.

When the docstring was duplicated the first docstring was documenting the second one so the function wasn't being documented at all. See https://github.com/JuliaLang/julia/issues/12737.

malmaud committed:
Merge pull request #13608 from JuliaLang/revert-13604-patch-1

Revert "Update types.rst"

artkuo commented:
That was indeed the error. I found out that during the build process, any docstrings in files compiled before operators.jl (line 11) would cause this error. Line 11 reads: `==(x,y) = x === y` which I believe is addressed by #13602. And of course, ranges.jl is one of about a dozen files that come before operators.jl.

stevengj commented:
Probably best to wait for 13602 to be merged, and then put the docstring back before the function, and finally [squash the commits](http://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git).  Then I think it should be ready to merge.

tkelman committed:
Merge pull request #13602 from MichaelHatherly/mh/doc-bootstrap-fix

Fix doc bootstrap method error.

andreasnoack committed:
Merge pull request #13496 from JuliaLang/anj/ldiv

Deprecate A_ldiv_B!(SparseMatrixCSC, StridedVecOrMat) thereby fixing #10787

jakebolewski committed:
Merge pull request #13514 from polarke/fix-sharedarray-constructors

Make the constructors of Array and SharedArray consistent

artkuo committed:
hi

artkuo commented:
Had to pull #13602 to test, but didn't know how to push back to this branch without including all the commits that had happened upstream. Closing this and starting a new one that should be clean.

KristofferC commented:
You can do `git rebase master` from your branch and it tries to put your commits on top of the newest commit on master.

artkuo commented:
Please see #13615.
